doc/specification.md

```diff
@@ -0,0 +1,624 @@
+RPC session.
+
+*   `DELETE` - When a client disconnects, the device should deletes all gRIBI
+    entries, received from that client, in RIB and FIB.
+*   `PRESERVE` - A client's disconnection SHOULD NOT trigger the device to
+    delete any gRIBI entry, received from that client, in RIB or FIB.
+
+No matter which mode the `Modify` RPC session is operating in, it is always the
+new primary client's (in case of [`SINGLE_PRIMARY`](\(#414-redundancy-mode\)))
+or other clients' (in case of [`ALL_PRIMARY`](#414-redundancy-mode))
+responsibility to do the reconciliation (e.g. via [`Get`](#42-get-rpc) and
+[`Modify`](#41-modify-rpc) RPC).
+
+### 4.1.6 Acknowledgement Mode
+
+The acknowledgement mode indicates when a response should be sent to a client as
+the result of executing the received AFT operations.
+[4.1.1](#411-client-server-session-negotiation) defines how the mode is agreed
+between client and server.
+
+`Modify` can operate in one of the following acknowledgement modes:
+
+*   `RIB_ACK`: After sending an AFT operation, the client expects the device to
+    respond whether if the AFT operation has been successfully programmed in the
+    RIB.
+*   `RIB_AND_FIB_ACK`: After sending an AFT operation, the client expects the
+    device to respond whether if the AFT operation has been successfully
+    programmed in both RIB and FIB.
+
+The response is reflected in `AFTResult.status` (see
+[4.1.3.2](#4132-aft-operation-response)).
+
+### 4.1.7 About gRIBI Server Caching
+
+gRIBI server implementation is not required to cache entries that have been
+removed, or invalidated by the system.
+
+The implication of this is that:
+
+*   When a VRF is removed (e.g. accidentally by user via cli):
+    *   The device is not required to maintain gRIBI objects in the FIB or RIB.
+    *   New `Get`([4.2](#42-get-rpc)) or `Flush`([4.3](#43-flush-rpc)) should
+        return corresponding errors (because the VRF is no longer there)
+    *   When the VRF is added back, the server is not required to restore all
+        the gRIBI objects without an explicit client request.
+    *   Since these external actions (e.g., configuration changes) can impact
+        whether a particular entry is being used by the server, the client MAY
+        monitor the state of the selected entries via subscribing to the
+        Abstract Forwarding Table (AFT) information via gNMI Subscribe.
+
+### 4.1.8 gRIBI Route Preference
+
+A device might learn routing information of the same destination from different
+protocols (e.g., static route, gRIBI, OSPF, BGP, etc.), and prefer one over the
+others. The preference is often configurable and indicated by different values
+(often known as Administrative Distance or Route Preference) in a network device
+OS. The values are locally significant to different device OS.
+
+This spec does not enforce the exact value that a device OS should assign to
+gRIBI protocol. However, the value MUST be configurable or by default between
```

Comment:
Is this quite right? I think this changed meaning in this edit. Particularly, my interpretation of the previous intent was it:

 * MUST default to between static and everything else.
 * MAY be configurable

whereas now we say that it:

 * MUST be configurable
 * OR/MAY have a value between static and everything else

I prefer the former to the latter.

Reply:
Sorry, forgot to call out this edit (and this is the only edit that changed meaning). I can't not tell for sure which one is better. 

IMHO`MUST default to between static and everything else.` + `MAYBE configurable` sounds too strong to me that excludes the future possibility from other protocols. 

Current wording is `MUST be configurable` **OR** `default to between static and everything else` might be better because: 
* pretty much all vendors support configurable route preference for different protocols anyway. 
* more flexible to other possible routing protocols in the future.

Thoughts?

Reply:
In the latter case, I think just saying it `MUST be configurable` sounds best since this is a strict super-set of the other behaviour of being less-preferable-than-static-more-preferable-than-everything-else. The cost that comes with is obviously having the implementation for the configurability, and needing to have the configuration explicitly specify the preference. Those seem reasonable to me.

Reply:
I"m ok with that. Updated. Thanks!